Assn.1 Refactoring To-do App used to Redux

TaeWoongKim2 opened the pull request:
"**To-do 리스트 Redux를 사용하여 리팩터링 하기**"

## **과제 목표**

상태 관리를 `Redux`로 변경해도 모든 컴포넌트 테스트가 통과해야 합니다. 추가로 `reducer`들에 대한 테스트를 추가해 주세요.

## **제한 조건**

모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 먼저 테스트를 추가하고 commit 한 뒤 실제 기능을 추가하고 commit합니다. commit 메시지 로그는 아래와 같이 남아야 합니다.

> Restaurant에 대한 테스트를 추가한다
> Restaurant을 구현한다
> Restaurant를 리팩토링한다

## **수행한 작업**

- [x] Clone the assn1 project and listen up the lecture 1 about Redux.

## **해야할 작업**

- [x] 나머지 Redux 강의 듣고 따라하기.
- [ ] Redux 공식 문서 보기.
- [ ] Flux Architecture 공식 문서 보기.

TaeWoongKim2 commented:
RP을 잘못 올려서 이리저리 시도 중 원복시켰습니다..!!
Commit 메시지가 TEST로 단순한 것은 스킵해주세요 ㅠㅠ.. 😅

TaeWoongKim2 commented:
## **수행한 작업**

- [x] Clone the Assn.1 named 'Refactoring To-do App used to Redux'.
- [x] 코드숨 Redux 강의 듣고 따라하기.


## **해야할 작업**

- [ ] Redux 공식 문서 보기.
- [ ] Flux Architecture 공식 문서 보기.
- [ ] **Assn.1 'Refactoring To-do App used to Redux' 원복 후 직접 리덕스 리팩토링하기.**

---

Redux 및 Flux에 대해 공부중입니다!!!

리덕스에 대한 감을 잡기 위해 강의를 듣고 따라서 코드를 구현해 봤습니다.

이후, 코드를 초기화 후 혼자서 다시 구현해볼 예정입니다!!!

wholemann commented:
![Screenshot](https://user-images.githubusercontent.com/47380072/96212454-eb863000-0fb1-11eb-95e5-8acc821f3482.png)

TDD로 진행해주세요~

TaeWoongKim2 committed:
Try Again Alone

git reset --hard "버전 id"

TaeWoongKim2 committed:
App 컴포넌트 리팩토링

리덕스를 적용하기 위해 To-do App 상태에 대한 관심사 분리.

TaeWoongKim2 committed:
[Refactoring] index에 Provider 추가.

store를 App 컴포넌트에 사용하기 위해 Provider 추가.

TaeWoongKim2 committed:
[Refactor] App  상태 처리 관심사 reducer로 분리

App 컴포넌트의 상태를 처리하는 부분을 store.js의 reducer로 분리.

TaeWoongKim2 committed:
[Refactor] App에 대한 Action creator 분리

App 컴포넌트에서 action creator를 담당하는 부분을 actions.js로 분리.

TaeWoongKim2 committed:
[Refactor/Green] App에 대한 테스트 리팩토링

1st, 테스트용 가상 store 생성을 위해 __mock__ 생성.
2nd, jest.mock 및 useSelector.mockImplementation를 통한 가상 상태 생성.

TaeWoongKim2 committed:
[Refactor] store에서 reducer를 분리

Make a js file name reduce for divide store.

TaeWoongKim2 committed:
[Red] reducer에 대한 테스트 구현

1st, updateTaskTitle에 대한 테스트 구현.
2nd, addTask에 대한 테스트 구현.
3rd, deleteTask에 대한 테스트 구현.

TaeWoongKim2 committed:
[Green] reducer 테스트에 맞춰 오류 수정

reducer 테스트 통과를 위한 코드 구현:
If task title is empty when click the add button, nothing happens!

TaeWoongKim2 committed:
[Red] ListContainer에 대한 테스트

ListContainer 컴포넌트에 대한 테스트 작성.

TaeWoongKim2 committed:
[Green] ListContainer 구현

ListContainer 테스트에 따른 ListContainer 구현.

TaeWoongKim2 committed:
[Refactor] ListContainer 오타 수정

function App() -> function ListContainer().

TaeWoongKim2 committed:
[Red] InputContainer에 대한 테스트

InputContainer에 대한 테스트 시나리오 작성.

TaeWoongKim2 committed:
[Red/Green] InputContainer 테스트 수정 및 구현

1st, getByText(/아무것도 안 하기/) -> getByDisplayValue(/아무것도 안 하기/)
2nd, InputContainer 구현.

TaeWoongKim2 committed:
[Refactor] InputContainer

Modify, function App() -> function InputContainer()

TaeWoongKim2 committed:
[Refactor] Page 컴포넌트 단순화

1st, Page 컴포넌트 InputContainer and ListContainer로 변경.
2nd, App 컴포넌트 단순화.

TaeWoongKim2 committed:
[Refactor] Page 컴포넌트 삭제

1st, App 컴포넌트로 InputContainer 및 ListContainer 병합.
2nd, App 테스트에 테스트 추가.

TaeWoongKim2 committed:
[Refactor] InputContainer에 대한 테스트 리팩토링

handleChangeTitle() 에 대한 action creator 테스트 추가.

TaeWoongKim2 committed:
[Refactor] ListContainer에 대한 테스트 리팩토링

handleClickDeleteTask() 에 대한 테스트 추가.
Coverage 100% 달성.

TaeWoongKim2 committed:
[Refactor] reducer coverage 100%

action type 이 존재하지 않을 경우 테스트 추가.
reducer.js 에 default param 추가 구현.

TaeWoongKim2 committed:
[Refactor] reducer action type null에 대한 코드 수정

reducer.js action type default param 에 대한 조건문 수정

wholemann commented on src/InputContainer.test.jsx:
```diff
@@ -6,28 +6,53 @@ import { useDispatch, useSelector } from 'react-redux';
     expect(getByDisplayValue(/아무것도 안 하기/)).not.toBeNull();
+  });
+
+  it('changes a task title', () => {
```
새로운 테스트 케이스가 추가됐군요.

사소하지만 이건 refactoring이라고 보긴 좀 어렵고 새로운 테스트 케이스 추가입니다ㅎㅎ RED가 뜨겠죠?

TaeWoongKim2 replied:
새로운 테스트가 추가된 것은 RED로 하면 되는군요!
참고 하겠습니다 :)

wholemann replied:
TDD 2개의 간단한 룰은 정말 간단하죠.

- 먼저 자동화된 테스트에서 실패하지 않는 한 새로운 코드를 작성하지 않는다.
- 중복을 제거한다.
